### DISCUSSION TITLE
[@types/aws-lambda]: AWS Lambda Integration with API Gateway 502 errors

### DISCUSSION THREAD

**Discussion by sterankin** [DefinitelyTyped/DefinitelyTyped#71065] [Issues with a @types package] [answered]
I have 2 simple `GET` requests in an API Gateway with Lambda Integrations.  Both are simple Hello World type examples and the lambda code is almost identical - with the exception that one Lambda is `Async` and the other is not.

The Async one works _fine_, the non-Async call errors at the Api Gateway with a 502 error:

```
    {
      "message": "Internal server error"
    }
    
    x-amzn-errortype	InternalServerErrorException
```

**I am lost as to why this could be.**

I have tracing and logging enabled in the API Gateway and I can see the logs which show the response as malformed:

> 
>     Received response. Status: 200, 
>     Integration latency: 25 ms 
>     Method completed with status: 502 
>     Execution failed due to configuration error: Malformed Lambda proxy response 
>     Endpoint response body before transformations: null

Here is the code for both lambdas.  The response is identical in each. Both lambdas have the same roles and permissions etc.  Both response conform to the AWS required [response format][1].  Both Lambdas execute ok without error.  The error is at the API Gateway side.

What could be causing this issue?

// works ok

```
    import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
    
    export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
      console.log('Received event:', JSON.stringify(event, null, 2));
      // Example of a fake async operation to appease ESLint
      await new Promise((resolve) => setTimeout(resolve, 0));
    
      const response: APIGatewayProxyResult = {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          message: 'Hello, World!',
          input: event,
        }),
      };
    
      return response;
    };

```

// does not work:

```
    import { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
    
    export function handler(event: APIGatewayProxyEvent, context: Context): APIGatewayProxyResult {
      console.log('event', event);
      console.log('context', context);
    
      const response: APIGatewayProxyResult = {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          message: 'Hello, World!',
          input: event,
        }),
      };
    
      return response;
    }
```

using this version:

```
  "dependencies": {
    "@types/aws-lambda": "^8.10.72",
    "aws-cdk-lib": "~2.158.0",
    "source-map-support": "^0.5.9"
  },
```

  [1]: https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format

[a comment by an automated account is not shown]

**Comment by jamesorlakin** [comment, marked as answer]
The issue is your second example isn't a compliant Lambda handler. See https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-callback

The jist is that the only thing a Lambda handler can return is a Promise. This library only provides types and thus makes no difference at runtime. If you'd like to avoid Promises and/or async/await then you can use the older callback mechanism which is given as an additional parameter to your handler function. E.g.

```js
import { APIGatewayProxyEvent, APIGatewayProxyResult, Context, Callback } from 'aws-lambda';

export function handler(event: APIGatewayProxyEvent, context: Context, callback: Callback<APIGatewayProxyResult>) {
  console.log('event', event);
  console.log('context', context);

  const response: APIGatewayProxyResult = {
    statusCode: 200,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      message: 'Hello, World!',
      input: event,
    }),
  };

  callback(null, response);
}
```

But unless you have a really good reason just keep it simple and use async + await would be my advice.
